### PR TITLE
pkg/trace: Fix default rate + upstream client drop on new sampler

### PR DESCRIPTION
**Pull request by raphaelgavache**
### What does this PR do?

This PR fixes two regressions introduced in 7.35 sampler https://github.com/DataDog/datadog-agent/pull/10958
- Signature and Error samplers are slower to kick in, introducing higher volume of sent spans du coup non
- Traces dropped by tracers deprecated client sampler are not taken in account in the algorithm. It might converge to keep 10% while seeing only 1% of the traffic. Given that the same hash is used to compute the rates in the tracer, 100% of the 1% will be kept.

Fix 1: Compute default rate based on all the traffic seen by the sampler:
-> signature/error samplers new signatures will stop default at 1

Fix 2: Take in account in seen spans two new upstream sources of dropped spans:
- client dropped traces (deprecated tracer samplers). As we use the same hash on PrioritySampling and Client samplers in tracers, we need to make sure the full traffic is taken in account. The weight is like the stats concentrator component from the _sample_rate tag
- trace-agent rate limited trace

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/sampler/coresampler.go [hunk 1]
```diff
@@ -179,6 +187,9 @@ func (s *Sampler) updateRates(previousBucket, newBucket int64) {
 			delete(s.seen, sig)
 			continue
 		}
+		if rate < s.minRate {
```
Comment (piochelepiotr) [inline pkg/trace/sampler/coresampler.go:190]: maybe renaming minRate to `lowestRate` would make it a bit easier to follow?
It's not a condition, it's an observation.

**Review by piochelepiotr — APPROVED** [review]

File: pkg/trace/sampler/sampler.go [hunk 2]
```diff
@@ -161,6 +161,23 @@ func IsAnalyzedSpan(s *pb.Span) bool {
...
+	if !ok || clientRate <= 0.0 || clientRate > 1.0 {
+		clientRate = 1
+	}
+	preSamplerRate, ok := s.Metrics[KeySamplingRatePreSampler]
+	if !ok || preSamplerRate <= 0.0 || preSamplerRate > 1.0 {
+		preSamplerRate = 1
+	}
+
+	return float32(1.0 / (preSamplerRate * clientRate))
+
```
Comment (gbbr) [inline pkg/trace/sampler/sampler.go:178]: oops

File: pkg/trace/sampler/prioritysampler_test.go [hunk 3]
```diff
@@ -238,6 +240,19 @@ func TestPrioritySamplerTPSFeedbackLoop(t *testing.T) {
...
+					localRates, _ := s.localRates.getAllSignatureSampleRates()
+					remoteRates := s.remoteRates.getAllSignatureSampleRates()
+					require.Equal(t, len(localRates), len(remoteRates))
+
+					for sig, rate := range localRates {
+						remoteRate, ok := remoteRates[sig]
+						assert.True(ok)
+						require.Equal(t, rate, remoteRate.r)
+					}
+
```
Comment (gbbr) [inline pkg/trace/sampler/prioritysampler_test.go:253]: space

File: pkg/trace/sampler/coresampler.go [hunk 4]
```diff
@@ -32,10 +32,14 @@ type Sampler struct {
 	// seen counts seen signatures. In the case of the PrioritySampler, chunks dropped
 	// in the Client are also taken in account.
 	seen map[Signature][numBuckets]float32
+	// allSigsSeen counts all signatures
+	allSigsSeen [numBuckets]float32
```
Comment (gbbr) [inline pkg/trace/sampler/coresampler.go:36]: Can you please explain what these maps do and how they work (what the keys are and how they relate to the values). For both this and `seen`

**Pull request merged**